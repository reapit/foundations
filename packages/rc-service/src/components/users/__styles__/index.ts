import { css } from '@linaria/core'
import { styled } from '@linaria/react'

export const ControlsContainer = styled.div`
  background-color: #fff;
  margin-bottom: 0.5rem;
  width: 100%;
`

export const inputFullWidth = css`
  input {
    width: 100%;
  }
`

export const DisplayChip = styled.span`
  background: var(--color-grey-light);
  color: var(--color-grey-dark);
  font-size: var(--font-size-small);
  border-radius: 1rem;
  padding: 0.25rem 0.625rem;
  margin: 0.25rem 0.5rem 0.25rem 0;
  display: inline-block;
`

export const twoColTable = css`
  grid-template-columns: repeat(2, 1fr);
`
