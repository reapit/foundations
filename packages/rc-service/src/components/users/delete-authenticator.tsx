import React, { FC, useEffect } from 'react'
import { Button } from '@reapit/elements'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { SendFunction, UpdateActionNames, updateActions, useReapitUpdate } from '@reapit/use-reapit-data'

interface DeleteAuthenticatorProps {
  authenticatorId?: string
  refreshAuthenticators: () => void
  userId?: string
}

export const handleRefresh = (refreshAuthenticators: () => void, hasDeleted?: boolean) => () => {
  if (hasDeleted) {
    refreshAuthenticators()
  }
}

export const handleDeleteAuthenticator = (deleteAuthenticator: SendFunction<void, boolean>) => () => {
  deleteAuthenticator()
}

export const DeleteAuthenticator: FC<DeleteAuthenticatorProps> = ({
  authenticatorId,
  userId,
  refreshAuthenticators,
}) => {
  const [deleteAuthenticatorLoading, , deleteAuthenticator, hasDeleted] = useReapitUpdate<void, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.deleteUserAuthenticator],
    method: 'DELETE',
    uriParams: {
      userId,
      authenticatorId,
    },
  })
  useEffect(handleRefresh(refreshAuthenticators, hasDeleted), [hasDeleted])

  return (
    <Button
      intent="primary"
      loading={deleteAuthenticatorLoading}
      disabled={deleteAuthenticatorLoading}
      onClick={handleDeleteAuthenticator(deleteAuthenticator)}
    >
      Reset Authenticator
    </Button>
  )
}
