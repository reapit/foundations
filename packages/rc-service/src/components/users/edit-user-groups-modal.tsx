import { useModal, Button, Modal, Icon } from '@reapit/elements'
import React, { FC } from 'react'
import { EditUserGroups, EditUserGroupsProps } from './edit-user-groups'

export const EditUserGroupsModal: FC<EditUserGroupsProps> = (props) => {
  const { modalIsOpen, openModal, closeModal } = useModal()

  return (
    <>
      <Button intent="neutral" onClick={openModal}>
        <Icon icon="edit" />
      </Button>
      <Modal
        title="Edit User Groups"
        isOpen={modalIsOpen}
        onModalClose={() => {
          closeModal()
        }}
      >
        <EditUserGroups {...props} closeModal={closeModal} />
      </Modal>
    </>
  )
}
