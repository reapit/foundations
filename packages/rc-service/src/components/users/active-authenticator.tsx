import React, { FC } from 'react'
import { elMb11, elFadeIn, Subtitle, Grid, Col, BodyText, Table } from '@reapit/elements'
import { cx } from '@linaria/core'
import { AuthenticatorModel } from '@reapit/foundations-ts-definitions'
import dayjs from 'dayjs'
import { DeleteAuthenticator } from './delete-authenticator'

interface ActiveAuthenticatorProps {
  activeAuthenticator: AuthenticatorModel
  refreshAuthenticators: () => void
}

export const ActiveAuthenticator: FC<ActiveAuthenticatorProps> = ({ activeAuthenticator, refreshAuthenticators }) => {
  const { id, type, created, status, modified, userId } = activeAuthenticator
  return (
    <Table
      rows={[
        {
          cells: [
            {
              label: 'Authenticator Type',
              value: type === 'sms' ? 'Mobile SMS' : 'Authenticator App',
            },
            {
              label: 'Status',
              value: status === 'inProgress' ? 'Currently Configuring' : status === 'active' ? 'Active' : 'Disabled',
            },
            {
              label: 'Created',
              value: created ? dayjs(created).format('DD/MM/YYYY HH:mm') : '-',
            },
            {
              label: 'Last Updated',
              value: modified ? dayjs(modified).format('DD/MM/YYYY HH:mm') : '-',
            },
            {
              label: 'Actions',
              children: (
                <DeleteAuthenticator
                  authenticatorId={id}
                  userId={userId}
                  refreshAuthenticators={refreshAuthenticators}
                />
              ),
            },
          ],
        },
      ]}
    />
  )
}
