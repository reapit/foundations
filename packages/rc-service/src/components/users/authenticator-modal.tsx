import { useModal, Button, Modal, PersistentNotification, Loader, ButtonGroup } from '@reapit/elements'
import { useReapitGet, getActions, GetActionNames } from '@reapit/use-reapit-data'
import React, { FC } from 'react'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { AuthenticatorModel } from '@reapit/foundations-ts-definitions'
import { DisplayChip } from './__styles__'
import { ActiveAuthenticator } from './active-authenticator'

export const AuthenticatorModal: FC<{ userId?: string }> = ({ userId }) => {
  const { modalIsOpen, openModal, closeModal } = useModal()

  const [authenticators, authenticatorsLoading, , refreshAuthenticators] = useReapitGet<AuthenticatorModel[]>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUserAuthenticators],
    uriParams: { userId },
    fetchWhenTrue: [userId, modalIsOpen],
  })
  const activeAuthenticator = authenticators?.find((authenticator) => authenticator.status === 'active')

  return (
    <>
      <Button intent="neutral" onClick={openModal}>
        Fetch
      </Button>
      <Modal
        title="MFA Authenticator"
        isOpen={modalIsOpen}
        onModalClose={() => {
          closeModal()
        }}
      >
        {activeAuthenticator && !authenticatorsLoading ? (
          <ActiveAuthenticator
            activeAuthenticator={activeAuthenticator}
            refreshAuthenticators={refreshAuthenticators}
          />
        ) : !authenticatorsLoading ? (
          <PersistentNotification isFullWidth isExpanded isInline intent="primary">
            No authenticators configured for this user.
          </PersistentNotification>
        ) : null}
        {authenticatorsLoading && modalIsOpen ? <Loader /> : null}
        <ButtonGroup alignment="right">
          <Button intent="default" onClick={closeModal}>
            Close
          </Button>
        </ButtonGroup>
      </Modal>
    </>
  )
}
