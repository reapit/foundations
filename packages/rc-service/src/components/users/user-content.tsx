import React, { ChangeEvent, Dispatch, FC, SetStateAction, useState } from 'react'
import ErrorBoundary from '../error-boundary'
import {
  BodyText,
  Button,
  ButtonGroup,
  Col,
  ElToggleItem,
  Grid,
  Loader,
  PersistantNotification,
  Subtitle,
  Table,
  TableCell,
  TableHeader,
  TableHeadersRow,
  TableRow,
  Toggle,
  elMb7,
  useModal,
} from '@reapit/elements'
import { DisplayChip, twoColTable } from './__styles__'
import {
  GetActionNames,
  SendFunction,
  UpdateActionNames,
  getActions,
  updateActions,
  useReapitGet,
  useReapitUpdate,
} from '@reapit/use-reapit-data'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import {
  AuthenticatorModel,
  EmailSuppressionModel,
  GroupModelPagedResult,
  UpdateUserModel,
  UserInfoModel,
  UserModel,
} from '@reapit/foundations-ts-definitions'
import { useReapitConnect } from '@reapit/connect-session'
import { getIsAdmin } from '../../utils/is-admin'
import { ActiveAuthenticator } from './active-authenticator'
import { EditUserGroups } from './edit-user-groups'

export interface UserContentProps {
  user: UserModel
  userGroups: GroupModelPagedResult | null
  refreshUsers: () => void
}

export interface ShouldFetchState {
  authenticators?: boolean
  password?: boolean
  suppressionList?: boolean
  officeGroups?: boolean
  loginLogs?: boolean
  userGroups?: boolean
}

export const handleUpdateUserStatus =
  (updateUser: SendFunction<UpdateUserModel, boolean>, user: UserModel, refreshUsers: () => void) =>
  async (event: ChangeEvent<HTMLInputElement>) => {
    const isChecked = event.target.checked
    const isUpdated = await updateUser({ ...(user as UpdateUserModel), inactive: !isChecked })
    if (isUpdated) {
      refreshUsers()
    }
  }

export const handleShouldFetch =
  (setShouldFetch: Dispatch<SetStateAction<ShouldFetchState>>, shouldFetchState: ShouldFetchState) => () => {
    setShouldFetch(shouldFetchState)
  }

export const handleResetPassword =
  (
    deleteUserPassword: SendFunction<void, boolean>,
    refreshUsers: () => void,
    setShouldFetch: Dispatch<SetStateAction<ShouldFetchState>>,
  ) =>
  async () => {
    const hasReset = await deleteUserPassword()

    if (hasReset) {
      refreshUsers()
      setShouldFetch({})
    }
  }

export const handleEmailSuppressionList =
  (
    removeFromSuppressionList: SendFunction<void, boolean>,
    refreshUsers: () => void,
    setShouldFetch: Dispatch<SetStateAction<ShouldFetchState>>,
  ) =>
  async () => {
    const hasBeenRemoved = await removeFromSuppressionList()

    if (hasBeenRemoved) {
      refreshUsers()
      setShouldFetch({})
    }
  }

export const UserContent: FC<UserContentProps> = ({ user, refreshUsers, userGroups }) => {
  const [shouldFetch, setShouldFetch] = useState<ShouldFetchState>({})
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { isSupport, isFoundationsAdmin } = getIsAdmin(connectSession)
  const { Modal } = useModal()

  const [authenticators, authenticatorsLoading, , refreshAuthenticators] = useReapitGet<AuthenticatorModel[]>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUserAuthenticators],
    uriParams: { userId: user.id },
    fetchWhenTrue: [user.id, shouldFetch.authenticators],
  })

  const [userInfo, userInfoLoading] = useReapitGet<UserInfoModel>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUserInfo],
    queryParams: { email: encodeURIComponent(user.email ?? ''), includeIdpData: true },
    fetchWhenTrue: [user.email, shouldFetch.officeGroups || shouldFetch.loginLogs],
  })

  const [userSuppressionList, userSuppressionListLoading] = useReapitGet<EmailSuppressionModel>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUserSuppressionList],
    uriParams: { userId: user.id },
    fetchWhenTrue: [user.id, shouldFetch.suppressionList],
    onError: () => {
      // suppress error
    },
  })

  const [userUpdateLoading, , updateUser] = useReapitUpdate<UpdateUserModel, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.updateUser],
    method: 'PUT',
    uriParams: {
      userId: user.id,
    },
  })

  const [userPasswordLoading, , deleteUserPassword] = useReapitUpdate<void, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.deleteUserPassword],
    method: 'DELETE',
    uriParams: {
      userId: user.id,
    },
  })

  const [userSuppressionListDeleteLoading, , deleteUserSuppressionList] = useReapitUpdate<void, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.deleteUserSuppressionList],
    method: 'DELETE',
    uriParams: {
      userId: user.id,
    },
  })

  const activeAuthenticator = authenticators?.find((authenticator) => authenticator.status === 'active')
  const isLoading =
    userUpdateLoading ||
    authenticatorsLoading ||
    userPasswordLoading ||
    userSuppressionListLoading ||
    userSuppressionListDeleteLoading ||
    userInfoLoading

  return (
    <ErrorBoundary>
      <Grid className={elMb7}>
        <Col>
          <Subtitle>Groups</Subtitle>
          <BodyText hasGreyText>
            {user?.groups?.sort().map((group, index) => (
              <DisplayChip key={index}>{group!}</DisplayChip>
            ))}
          </BodyText>
        </Col>
        <Col>
          <Subtitle>User Claims</Subtitle>
          <BodyText hasGreyText>
            {user.userClaims?.map(({ claim, value }, index) => {
              if (claim?.includes('payrollNumber')) return null
              return (
                <DisplayChip key={index}>
                  {claim}: {value}
                </DisplayChip>
              )
            })}
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Organisation Claims</Subtitle>
          <BodyText>
            {user.organisationClaims?.map(({ claim, value }, index) => {
              if (claim?.includes('EmailHeader')) return null
              return (
                <DisplayChip key={index}>
                  {claim}: {value}
                </DisplayChip>
              )
            })}
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Status</Subtitle>
          <BodyText hasGreyText>
            {isSupport ? (
              <Toggle
                id={`active-toggle-${user.id}`}
                defaultChecked={!user.inactive}
                onChange={handleUpdateUserStatus(updateUser, user, refreshUsers)}
              >
                <ElToggleItem>Active</ElToggleItem>
                <ElToggleItem>Inactive</ElToggleItem>
              </Toggle>
            ) : user.inactive ? (
              'Inactive'
            ) : (
              'Active'
            )}
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Password</Subtitle>
          <BodyText hasGreyText>
            <Button intent="primary" onClick={handleShouldFetch(setShouldFetch, { password: true })}>
              Reset Password
            </Button>
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Suppression List</Subtitle>
          <BodyText hasGreyText>
            <Button
              intent="primary"
              loading={userSuppressionListLoading}
              disabled={!isSupport || userSuppressionListLoading}
              onClick={handleShouldFetch(setShouldFetch, { suppressionList: true })}
            >
              Check
            </Button>
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Office Groups</Subtitle>
          <BodyText hasGreyText>
            <Button
              loading={userInfoLoading && shouldFetch.officeGroups}
              intent="primary"
              disabled={userInfoLoading && shouldFetch.officeGroups}
              onClick={handleShouldFetch(setShouldFetch, { officeGroups: true })}
            >
              Check
            </Button>
          </BodyText>
        </Col>
        <Col>
          <Subtitle>Login Info</Subtitle>
          <BodyText hasGreyText>
            <Button
              loading={userInfoLoading && shouldFetch.loginLogs}
              intent="primary"
              disabled={!isSupport || (userInfoLoading && shouldFetch.loginLogs)}
              onClick={handleShouldFetch(setShouldFetch, { loginLogs: true })}
            >
              Check
            </Button>
          </BodyText>
        </Col>
        <Col>
          <Subtitle>MFA Authenticators</Subtitle>
          <ButtonGroup alignment="left">
            <Button
              intent="primary"
              loading={authenticatorsLoading}
              disabled={!isSupport || Boolean(authenticatorsLoading || authenticators)}
              onClick={handleShouldFetch(setShouldFetch, { authenticators: true })}
            >
              Fetch Current Authenticators
            </Button>
          </ButtonGroup>
        </Col>
        <Col>
          <Subtitle>Manage User Groups</Subtitle>
          <ButtonGroup alignment="left">
            <Button
              intent="primary"
              disabled={(!isSupport && !isFoundationsAdmin) || !userGroups}
              onClick={handleShouldFetch(setShouldFetch, { userGroups: true })}
            >
              Load Group Management
            </Button>
          </ButtonGroup>
        </Col>
      </Grid>
      {isLoading && <Loader />}
      {activeAuthenticator && isSupport && shouldFetch.authenticators ? (
        <ActiveAuthenticator activeAuthenticator={activeAuthenticator} refreshAuthenticators={refreshAuthenticators} />
      ) : shouldFetch.authenticators ? (
        <PersistantNotification isFullWidth isExpanded isInline intent="primary">
          No authenticators configured for this user.
        </PersistantNotification>
      ) : null}
      {userInfo && shouldFetch.officeGroups && userInfo.officeGroupIds ? (
        <>
          <Table>
            <TableHeadersRow className={twoColTable}>
              <TableHeader>Group Name</TableHeader>
              <TableHeader>Office Ids</TableHeader>
            </TableHeadersRow>
            <TableRow className={twoColTable}>
              <TableCell>{userInfo.officeGroupName}</TableCell>
              <TableCell>{userInfo.officeGroupIds}</TableCell>
            </TableRow>
          </Table>
        </>
      ) : shouldFetch.officeGroups ? (
        <PersistantNotification isFullWidth isExpanded isInline intent="primary">
          User not part of an office group.
        </PersistantNotification>
      ) : null}
      {userInfo && shouldFetch.loginLogs && userInfo.idpData?.authEvents?.length ? (
        userInfo.idpData?.authEvents.slice(0, 5).map((event) => <DisplayChip key={event}>{event}</DisplayChip>)
      ) : shouldFetch.loginLogs ? (
        <PersistantNotification isFullWidth isExpanded isInline intent="primary">
          No info available for this user.
        </PersistantNotification>
      ) : null}
      {shouldFetch.userGroups && <EditUserGroups refreshUsers={refreshUsers} user={user} userGroups={userGroups} />}
      {shouldFetch.password && (
        <Modal title="Confirm reset user password" isOpen={Boolean(shouldFetch.password)}>
          <BodyText>Are you sure you wish to reset this user password?</BodyText>
          <ButtonGroup alignment="center">
            <Button
              intent="danger"
              disabled={userPasswordLoading}
              onClick={handleResetPassword(deleteUserPassword, refreshUsers, setShouldFetch)}
            >
              Reset Password
            </Button>
            <Button intent="default" onClick={handleShouldFetch(setShouldFetch, {})}>
              Cancel
            </Button>
          </ButtonGroup>
        </Modal>
      )}
      {shouldFetch.suppressionList && !userSuppressionListLoading && (
        <Modal title="User Email Suppression List" isOpen={Boolean(shouldFetch.suppressionList)}>
          {userSuppressionList ? (
            <BodyText>User is on the email suppression list - you can remove them below</BodyText>
          ) : (
            <BodyText>User is not on the email suppression list</BodyText>
          )}
          <ButtonGroup alignment="center">
            {userSuppressionList && (
              <Button
                intent="primary"
                disabled={userSuppressionListDeleteLoading}
                onClick={handleEmailSuppressionList(deleteUserSuppressionList, refreshUsers, setShouldFetch)}
              >
                Remove from list
              </Button>
            )}
            <Button intent="default" onClick={handleShouldFetch(setShouldFetch, {})}>
              Cancel
            </Button>
          </ButtonGroup>
        </Modal>
      )}
    </ErrorBoundary>
  )
}
