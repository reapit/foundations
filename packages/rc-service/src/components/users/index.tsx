import React, { Dispatch, FC, SetStateAction, useEffect, useState } from 'react'
import {
  Icon,
  Loader,
  Table,
  Pagination,
  PersistentNotification,
  FormLayout,
  InputGroup,
  InputWrap,
  elFadeIn,
  elMb11,
  Label,
  ToggleRadio,
  Select,
  BodyText,
} from '@reapit/elements'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { GroupModelPagedResult, UserModelPagedResult } from '@reapit/foundations-ts-definitions'
import ErrorBoundary from '../error-boundary'
import { useForm, UseFormWatch } from 'react-hook-form'
import { cx } from '@linaria/core'
import { GetActionNames, getActions, objectToQuery, useReapitGet } from '@reapit/use-reapit-data'
import debounce from 'just-debounce-it'
import dayjs from 'dayjs'
import { UserContent } from './user-content'

export interface UserFilters {
  email?: string
  name?: string
  organisationName?: string
  active?: string
  groupId?: string
  mfaEnabled?: string
}

export const handleSetAdminFilters =
  (setUserSearch: Dispatch<SetStateAction<UserFilters>>, watch: UseFormWatch<UserFilters>) => () => {
    const subscription = watch(debounce(setUserSearch, 200))
    return () => subscription.unsubscribe()
  }

export const UsersPage: FC = () => {
  const [userSearch, setUserSearch] = useState<UserFilters>({})
  const [pageNumber, setPageNumber] = useState<number>(1)
  const { register, watch } = useForm<UserFilters>({ mode: 'all' })
  const emailQuery = {
    email: userSearch.email ? encodeURIComponent(userSearch.email) : undefined,
  }
  const queryParams = objectToQuery({
    ...userSearch,
    ...emailQuery,
  })

  const [users, usersLoading, , refreshUsers] = useReapitGet<UserModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUsers],
    queryParams: { pageSize: 12, pageNumber, ...queryParams },
    fetchWhenTrue: [],
  })

  const [userGroups] = useReapitGet<GroupModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getUserGroups],
    queryParams: { pageSize: 100 },
    fetchWhenTrue: [],
  })

  useEffect(handleSetAdminFilters(setUserSearch, watch), [])

  return (
    <ErrorBoundary>
      <form>
        <FormLayout hasMargin>
          <InputWrap>
            <InputGroup
              {...register('organisationName')}
              type="search"
              icon="searchSystem"
              placeholder="Search by organisation name"
              label="Org Name"
            />
          </InputWrap>
          <InputWrap>
            <InputGroup
              {...register('email')}
              type="search"
              icon="searchSystem"
              placeholder="Search by email"
              label="User Email"
            />
          </InputWrap>
          <InputWrap>
            <InputGroup
              {...register('name')}
              type="search"
              icon="searchSystem"
              placeholder="Search by user name"
              label="User Name"
            />
          </InputWrap>
          <InputWrap>
            <InputGroup>
              <Select {...register('groupId')}>
                <option key="default-option" value={''}>
                  Please Select
                </option>
                {userGroups?._embedded?.map(({ id }) => (
                  <option key={id} value={id}>
                    {id}
                  </option>
                ))}
              </Select>
              <Label htmlFor="myId">Select User Group</Label>
            </InputGroup>
          </InputWrap>
          <InputWrap>
            <Label>User Active</Label>
            <ToggleRadio
              hasGreyBg
              {...register('active')}
              options={[
                {
                  id: 'usr-active-all',
                  value: '',
                  text: 'All',
                  isChecked: true,
                },
                {
                  id: 'usr-active-true',
                  value: 'true',
                  text: 'Active',
                  isChecked: false,
                },
                {
                  id: 'usr-active-false',
                  value: 'false',
                  text: 'Inactive',
                  isChecked: false,
                },
              ]}
            />
          </InputWrap>
          <InputWrap>
            <Label>MFA Enabled</Label>
            <ToggleRadio
              {...register('mfaEnabled')}
              hasGreyBg
              options={[
                {
                  id: 'mfa-enabled-all',
                  value: '',
                  text: 'All',
                  isChecked: true,
                },
                {
                  id: 'mfa-enabled-true',
                  value: 'true',
                  text: 'Enabled',
                  isChecked: false,
                },
                {
                  id: 'mfa-enabled-false',
                  value: 'false',
                  text: 'Not Configured',
                  isChecked: false,
                },
              ]}
            />
          </InputWrap>
        </FormLayout>
      </form>
      {usersLoading ? (
        <Loader />
      ) : users?._embedded?.length ? (
        <>
          <BodyText>Total Users: {users?.totalCount}</BodyText>
          <Table
            className={cx(elFadeIn, elMb11)}
            rows={users?._embedded?.map((user) => {
              const {
                name,
                email,
                created,
                jobTitle,
                inactive,
                organisationName,
                organisationId,
                agencyCloudNegotiatorId,
              } = user
              return {
                cells: [
                  {
                    label: 'Name',
                    value: name ?? '-',
                    icon: 'usernameSystem',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Email',
                    value: email ?? '-',
                    icon: 'emailSystem',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Date Created',
                    value: created ? dayjs(created).format('DD-MM-YYYY') : '-',
                    icon: 'calendarSystem',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Job Title',
                    value: jobTitle ?? '-',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Organisation',
                    value: organisationName ?? '-',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Customer Id',
                    value: organisationId ?? '-',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Neg Id',
                    value: agencyCloudNegotiatorId ?? '-',
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                  {
                    label: 'Active',
                    value: (
                      <Icon icon={inactive ? 'closeSystem' : 'checkSystem'} intent={inactive ? 'danger' : 'success'} />
                    ),
                    narrowTable: {
                      showLabel: true,
                    },
                  },
                ],
                expandableContent: {
                  content: <UserContent user={user} refreshUsers={refreshUsers} userGroups={userGroups} />,
                },
              }
            })}
          />
          <Pagination
            callback={setPageNumber}
            currentPage={pageNumber}
            numberPages={Math.ceil((users?.totalCount ?? 1) / 12)}
          />
        </>
      ) : (
        <PersistentNotification isFullWidth isExpanded intent="secondary" isInline>
          No users found based on your current search.
        </PersistentNotification>
      )}
    </ErrorBoundary>
  )
}

export default UsersPage
