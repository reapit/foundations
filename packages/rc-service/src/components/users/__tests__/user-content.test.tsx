import { UserContent } from '../user-content'
import React from 'react'
import { render } from '../../../tests/react-testing'
import { mockUserModelPagedResult } from '../../../tests/__stubs__/users'

describe('UserContent', () => {
  it('Should match snapshot', async () => {
    const wrapper = render(
      <UserContent
        user={{
          ...(mockUserModelPagedResult._embedded ? mockUserModelPagedResult._embedded[0] : {}),
        }}
        userGroups={{ _embedded: [] }}
        refreshUsers={() => {}}
      />,
    )
  })
})
