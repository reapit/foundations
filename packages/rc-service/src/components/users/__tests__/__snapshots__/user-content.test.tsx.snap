// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserContent Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mocked-styled-93 el-snack-holder"
      />
      <div
        class="sc-gEvEer fQSvkU"
      >
        <div
          class="mocked-styled-115 el-table"
          data-has-call-to-action="false"
          data-has-expandable-action="false"
          data-num-columns-excl-action-col="5"
          role="table"
        >
          <div
            class="mocked-styled-101 el-table-headers-row"
            role="row"
          >
            <div
              aria-label="Organisation"
              class="mocked-styled-102 el-table-header"
              role="columnheader"
            >
              Organisation
            </div>
            <div
              aria-label="Organisation Claims"
              class="mocked-styled-102 el-table-header"
              role="columnheader"
            >
              Organisation Claims
            </div>
            <div
              aria-label="User Claims"
              class="mocked-styled-102 el-table-header"
              role="columnheader"
            >
              User Claims
            </div>
            <div
              aria-label="Office Groups"
              class="mocked-styled-102 el-table-header"
              role="columnheader"
            >
              Office Groups
            </div>
          </div>
          <div
            class="mocked-styled-110 el-table-row-container"
            role="row"
          >
            <div
              class="mocked-styled-103 el-table-row"
            >
              <div
                class="el-table-cell-has-dark-text el-table-cell"
                role="cell"
              >
                <div
                  class="mocked-styled-111 el-table-cell-content"
                >
                  <div
                    style="display: flex; align-items: center; gap: 8px; flex-wrap: wrap;"
                  >
                    <span>
                      an org
                    </span>
                    <mock-styled.span
                      classname="sc-eqUAAy cqznCV"
                      type="customer"
                    >
                      customer
                    </mock-styled.span>
                    <mock-styled.span
                      classname="sc-eqUAAy gNlkwv"
                      type="developer"
                    >
                      developer
                    </mock-styled.span>
                  </div>
                </div>
              </div>
              <div
                class="el-table-cell-has-dark-text el-table-cell"
                role="cell"
              >
                <div
                  class="mocked-styled-111 el-table-cell-content"
                >
                  None
                </div>
              </div>
              <div
                class="el-table-cell-has-dark-text el-table-cell"
                role="cell"
              >
                <div
                  class="mocked-styled-111 el-table-cell-content"
                >
                  None
                </div>
              </div>
              <div
                class="el-table-cell-has-dark-text el-table-cell"
                role="cell"
              >
                <div
                  class="mocked-styled-111 el-table-cell-content"
                >
                  <button
                    class="el-intent-neutral el-button"
                  >
                    <div
                      class="mocked-styled-0 el-button-loader"
                    />
                    Check
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mocked-styled-93 el-snack-holder"
    />
    <div
      class="sc-gEvEer fQSvkU"
    >
      <div
        class="mocked-styled-115 el-table"
        data-has-call-to-action="false"
        data-has-expandable-action="false"
        data-num-columns-excl-action-col="5"
        role="table"
      >
        <div
          class="mocked-styled-101 el-table-headers-row"
          role="row"
        >
          <div
            aria-label="Organisation"
            class="mocked-styled-102 el-table-header"
            role="columnheader"
          >
            Organisation
          </div>
          <div
            aria-label="Organisation Claims"
            class="mocked-styled-102 el-table-header"
            role="columnheader"
          >
            Organisation Claims
          </div>
          <div
            aria-label="User Claims"
            class="mocked-styled-102 el-table-header"
            role="columnheader"
          >
            User Claims
          </div>
          <div
            aria-label="Office Groups"
            class="mocked-styled-102 el-table-header"
            role="columnheader"
          >
            Office Groups
          </div>
        </div>
        <div
          class="mocked-styled-110 el-table-row-container"
          role="row"
        >
          <div
            class="mocked-styled-103 el-table-row"
          >
            <div
              class="el-table-cell-has-dark-text el-table-cell"
              role="cell"
            >
              <div
                class="mocked-styled-111 el-table-cell-content"
              >
                <div
                  style="display: flex; align-items: center; gap: 8px; flex-wrap: wrap;"
                >
                  <span>
                    an org
                  </span>
                  <mock-styled.span
                    classname="sc-eqUAAy cqznCV"
                    type="customer"
                  >
                    customer
                  </mock-styled.span>
                  <mock-styled.span
                    classname="sc-eqUAAy gNlkwv"
                    type="developer"
                  >
                    developer
                  </mock-styled.span>
                </div>
              </div>
            </div>
            <div
              class="el-table-cell-has-dark-text el-table-cell"
              role="cell"
            >
              <div
                class="mocked-styled-111 el-table-cell-content"
              >
                None
              </div>
            </div>
            <div
              class="el-table-cell-has-dark-text el-table-cell"
              role="cell"
            >
              <div
                class="mocked-styled-111 el-table-cell-content"
              >
                None
              </div>
            </div>
            <div
              class="el-table-cell-has-dark-text el-table-cell"
              role="cell"
            >
              <div
                class="mocked-styled-111 el-table-cell-content"
              >
                <button
                  class="el-intent-neutral el-button"
                >
                  <div
                    class="mocked-styled-0 el-button-loader"
                  />
                  Check
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
